text/0000-improved-ember-registry-apis.md

```diff
@@ -0,0 +1,556 @@
+
+-  registeredOption(fullName: string, optionName: string): RegisterOptions;
++  registeredOption(identifier: Identifier, optionName: string): RegisterOptions;
+
+-  registeredOptions(fullName: string): RegisterOptions;
++  registeredOptions(identifier: Identifier): RegisterOptions;
+
+-  registeredOptionsForType(type: string): RegisterOptions;
++  registeredOptionsForType(type: FactoryTypeIdentifier): RegisterOptions;
+
+-  resolveRegistration(type: string): Factory;
++  resolveRegistration(identifier: Identifier): Factory;
+
+-  unregister(fullName: string): void;
++  unregister(identifier: Identifier): void;
+ }
+```
+
+### Codemod
+
+All existing microsyntax invocations can be straightforwardly migrated to the new syntax with a codemod.
+
+### Deprecation messaging
+
+Invocation of v0 resolver APIs which require an `Identifier` will trigger the following deprecation message (using `lookup` as an example):
+
+> You invoked `Owner.lookup` with a `string` full name: `'<type>:<name>'`. This usage is deprecated and will be removed in Ember 4.0. Instead, pass an object identifier: `{ type: '<type>', name: '<name>' }`.
+
+Here, the message should substitute the *actual* passed type and name. For example, if the user tried to look up a service named `session` by invoking `lookup('service:session')` the deprecation will read:
+
+> You invoked `Owner.lookup` with a string full name: `'service:session'`. This usage is deprecated and will be removed in Ember 4.0. Instead, pass an identifier object: `{ type: 'service', name: 'session' }`.
+
+For methods which require a `FactoryTypeIdentifier`, the wording is adjusted appropriately:
+
+> You invoked `Owner.lookup` with a string type name: `'service'`. This usage is deprecated and will be removed in Ember 4.0. Instead, pass a factory type identifier object: `{ type: 'service' }`.
+
+Methods which support either `Identifier`s or `FactoryTypeIdentifier`s should display the appropriate variant of the message according to what the user actually supplied.
+
+### Rollout
+
+The rollout will be phased:
+
+- The new API will be introduced in a minor release as normal, *without* introducing the deprecation warning for the microsyntax-based design, and *including* the two codemods required for this API change: one for all Ember users, and one for TypeScript users (see [**Proposed Type Definitions**](#proposed-type-definitions) in the [**TypeScript Appendix**](#appendix-typescript) below.)
+
+- After *at least* one minor version, the deprecation message will be introduced. This will give addons time to adopt the new API *before* requiring 
+
+- We will supply a polyfill for the new behavior, supporting *at least* the current LTS. The behavior here can be polyfilled to many if not all existing LTS releases; we would likely limit this slightly, but would support *many* LTS releases.
+
+Changes to `ember-resolver` will be introduced in a major version, signaling clearly to users that there is a breaking change inbound.
```

Comment:
See note above, I don't actually think this needs a breaking change.

Reply:
Whoops, meant to update this one as well when I updated the other notes about back-compat! Thank you!

→ 923f508